### PR TITLE
[CUDA graphs] Make stream semantics of backward calls consistent with other cuda ops (ci-all edition)

### LINKED ISSUE #47028: torch.autograd.backward() fails to sync with other stream
The new test written for https://github.com/pytorch/pytorch/pull/45787 suggests a possible failure scenario, which indeed occurs.  It is a race condition, most often encountered by ROCm CI.

https://github.com/pytorch/pytorch/pull/45787/files#diff-893b1eea27352f336f4cd832919e48d721e4e90186e63400b8596db6b82e7450R1772-R1773

Putting torch.cuda.synchronize() right after backward() fixes the problem.

_Originally posted by @mcarilli in https://github.com/pytorch/pytorch/pull/45787#issuecomment-717417098_

### PR DESCRIPTION
**Pull request by mcarilli**
ci-all resubmit of https://github.com/pytorch/pytorch/pull/54227.

Tests look good except for a few distributed autograd failures (pytorch_linux_xenial_cuda10_2_cudnn7_py3_multigpu_test) and rocm failures (pr/pytorch-linux-bionic-rocm4.1-py3.6).

The common denominator in rocm failures appears to be multi-gpu activity: some [multiprocess DDP failures](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.1-py3.6-test1/8115/console), some [single-process failures](https://ci.pytorch.org/jenkins/job/pytorch-builds/job/pytorch-linux-bionic-rocm4.1-py3.6-test2/8115/console) where the single process has autograd ops that span devices. @jeffdaily @jithunnair-amd @sunway513, could one of you take a look? The streaming backward change is also beneficial to rocm, I expect.

For debugging rocm failures, I think we should ignore the multiprocess/DDP tests and focus on the single process cases. The root cause is probably the same and the single process cases are simpler.

----------------------------------

Update: Rocm failures are due to https://github.com/pytorch/pytorch/issues/59750. 
https://github.com/pytorch/pytorch/pull/57833/commits/2718a54032d0791ce90a9a95d15150c53727713e is a workaround, to be updated once https://github.com/pytorch/pytorch/issues/59750 is fixed.

### CODE REVIEW

File: torch/csrc/autograd/engine.cpp [hunk 1]
```diff
@@ -533,21 +542,62 @@ void GraphTask::exec_post_processing() {
...
+      cb_lock.unlock();
+      final_callbacks_[i]();
+      cb_lock.lock();
+    }
+  }
+
+  // For temporary BC, syncs default streams with caller_current_streams so callback results are also
+  // usable on user-facing default streams after backward()
+  for (const auto& caller_current_stream : caller_current_streams_filtered) {
+    const auto caller_default_stream = *caller_default_streams_[caller_current_stream.device_index()];
```
Comment (ngimel) [inline torch/csrc/autograd/engine.cpp:595]: You are dereferencing an optional without checking if it has value, this probably leads to test failure on ROCm

**Comment by ngimel** [comment]
ROCm and TensorPipe test failure look real, you might want to add more asserts to the code and resubmit (or maybe try to repro TensorPipe failure locally).

**Comment by ngimel** [comment]
@mcarilli can you please post what you have done for debugging? Now a lot of tests fail, so I don't even know which of the failures is the original. What are the asserts you've added, did they narrow things down?

**Comment by mcarilli** [comment]
@ngimel The original failing tests are distributed autograd tests
```
test_device_maps_backward_pass
test_gradients_synchronizations
test_dist_autograd_sync_streams
```
in pytorch_linux_xenial_cuda10_2_cudnn7_py3_multigpu_test
and several failures for tests with backward ops that span devices in pr/pytorch-linux-bionic-rocm4.2-py3.6.

My three debugging commits are the three labeled DEBUG. They are all targeted at the distributed autograd failures.
`comment out asserts in failing rpc tests` comments out some numerical checks in the distributed autograd tests to make sure the failure is due to an internal segfault rather than bad numerics.
`Comment out all stream syncs in exec_post_processing to isolate rpc segfault` comments out ALL my stream syncing logic in backward to see if the segfaults go away (they do). It also disables a bunch of non-distributed streaming backward tests, because I knew they'd start failing and I didn't want their failures to prevent the test from reaching the dist autograd tests.
`Restore some syncs, see if rpc segfaults return` comments some (not all) of the syncs back in, and the segfault returns. This narrows down lines that might be responsible for the segfault.

Since the PR is currently in a debugging state where some of the needed syncs are commented out, I'm not surprised to see some random failures. pytorch_linux_xenial_cuda10_2_cudnn7_py3_multigpu_test and pr/pytorch-linux-bionic-rocm4.2-py3.6 are the only tests I'm worried about. I think the others will go green again when I restore all the syncs.

File: torch/csrc/autograd/engine.cpp [hunk 2]
```diff
@@ -535,23 +544,65 @@ void GraphTask::exec_post_processing() {
...
+      // a c10::nullopt. I don't expect GraphTask's backward pass ran leaf nodes on
+      // any new devices, so the stashed streams should be enough.
+      // If leaf_stream.device_index() happens to be for a new device,
+      // operator* on the c10::nullopt will throw an error.
+      const auto caller_current_stream = *caller_current_streams_[leaf_stream.device_index()];
+
+      if (caller_current_stream != leaf_stream) {
+        auto event = c10::Event{c10::DeviceType::CUDA};
+        event.record(leaf_stream);
+        caller_current_stream.wait(event);
```
Comment (amathews-amd) [inline torch/csrc/autograd/engine.cpp:567]: Can you confirm that 
```
caller_current_stream.wait(event);
```
shouldnt be 
```
*caller_default_streams_[leaf_stream.device_index()].wait(event);
```
?

Comment (amathews-amd) [reply torch/csrc/autograd/engine.cpp:567]: or even: 
`
leaf_stream.wait(event);
`
I had the test passing on ROCm with this change.

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 3]
```diff
@@ -255,6 +262,12 @@ void DistEngine::computeDependencies(
     }
   }
 
+  if (might_use_cuda && will_use_cuda) {
+    // Collects current and default streams for devices where this process has a context,
+    // so GraphTask::exec_post_processing can sync them with leaf_streams.
+    task->stash_current_streams();
```
Comment (pritamdamania87) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp:268]: Shouldn't this be `graphTask`?

Comment (mcarilli) [reply torch/csrc/distributed/autograd/engine/dist_engine.cpp]: fixed

File: torch/csrc/autograd/engine.cpp [hunk 4]
```diff
@@ -533,24 +542,64 @@ void GraphTask::exec_post_processing() {
...
+    // final_callbacks run on the per-device caller_current_streams (the ambient streams
+    // surrounding the user's call to backward()). This has two benefits:
+    //  1. caller_current_streams have been synced with leaf_streams, so callbacks may
+    //     safely access any grad.
+    //  2. The callback's results can safely be used on (user-facing) caller_current_streams
+    //     after backward().
+    c10::MultiStreamGuard g(caller_current_streams_filtered);
+    // WARNING: Don't use a range-for loop here because more callbacks may be
+    // added in between callback calls, so iterators may become invalidated.
+    for (const auto i : c10::irange(final_callbacks_.size())) {
```
Comment (ngimel) [inline torch/csrc/autograd/engine.cpp:587]: please fix the range for loop here per warning

File: torch/csrc/autograd/engine.cpp [hunk 5]
```diff
@@ -895,6 +949,12 @@ auto Engine::compute_dependencies(Node* root, GraphTask& task, uint64_t min_topo
       }
     }
   }
+
+  if (might_use_cuda && will_use_cuda) {
```
Comment (ngimel) [inline torch/csrc/autograd/engine.cpp:953]:
```suggestion
  if (will_use_cuda) {
```

File: torch/csrc/distributed/autograd/engine/dist_engine.cpp [hunk 6]
```diff
@@ -255,6 +262,12 @@ void DistEngine::computeDependencies(
     }
   }
 
+  if (might_use_cuda && will_use_cuda) {
```
Comment (ngimel) [inline torch/csrc/distributed/autograd/engine/dist_engine.cpp:265]:
```suggestion
  if (will_use_cuda) {
```

**Pull request closed without merging**